hail/python/hail/expr/aggregators/aggregators.py

```diff
@@ -108,6 +109,102 @@ def make_seq_op(agg):
+                else:
+                    raise ExpressionException("'agg.explode' function cannot take scans, and 'scan.explode function cannot take aggregations'")
+            elif isinstance(ir, Ref) and ir.name == var:
+                raise ExpressionException("'agg.explode' can't reference collection element outside of aggregation")
+            else:
+                return ir.map_ir(rewrite)
+
+        return construct_expr(rewrite(aggregated._ir), aggregated.dtype, aggregated._indices, array_agg_expr._aggregations.push(Aggregation(aggregated)))
+
+
+    @typecheck_method(condition=expr_bool,
+                      aggregation=expr_any)
+    def filter(self, condition, aggregation):
+        if not condition._aggregations.empty():
+            raise ExpressionException("'agg.filter' does not support an already-aggregated expression as the argument to 'condition'")
+        if aggregation._aggregations.empty():
+            raise ExpressionException("'agg.filter' must take expression containing aggregation as argument to 'aggregation'")
+
+        unify_all(condition, aggregation)
+
+        def rewrite(ir):
+            if isinstance(ir, BaseApplyAggOp):
+                if self._as_scan == isinstance(ir, ApplyScanOp):
+                    old_seq_op = ir.a
+                    new_seq_op = If(condition._ir, old_seq_op, Begin([]))
+                    init_ops = ir.init_op_args if ir.init_op_args else []
+                    return ir.copy(new_seq_op, *ir.constructor_args, *init_ops)
+                else:
+                    raise ExpressionException("'agg.filter' function cannot take scans, and 'scan.filter function cannot take aggregations'")
+            else:
+                return ir.map_ir(rewrite)
+
+        new_ir = rewrite(aggregation._ir)
+        return construct_expr(new_ir, aggregation.dtype, aggregation._indices, condition._aggregations.push(Aggregation(condition, aggregation)))
+
+    def group_by(self, group, aggregation):
+        if not group._aggregations.empty():
+            raise ExpressionException("'agg.group_by' does not support an already-aggregated expression as the argument to 'group'")
+
+        if aggregation._aggregations.empty():
+            raise ExpressionException("'agg.group_by' must take expression containing aggregation as argument to 'aggregation'")
+
+        indices, aggregations = unify_all(group, aggregation)
+
+        agg_ops = {}
+
+        def rewrite_a(node):
+            if isinstance(ir, SeqOp):
+                return SeqOp(node.i, [group._ir] + node.args, new_agg_sig)
+            else:
+                return node.map_ir(rewrite_a)
+
+        def rewrite(ir):
+            if isinstance(ir, BaseApplyAggOp):
+                if self._as_scan == isinstance(ir, ApplyScanOp):
+                    var = Env.get_uid()
+                    typ = tint #FIXME: this might have to call into scala right now to get a type.
```

Comment:
I had to do that in my PR:

```
		                 new_agg_sig = seq_op[0].agg_sig
                 j_agg_sig = Env.hail().expr.Parser.parseAggSignature(str(new_agg_sig))

                 return_typ = HailType._from_java(Env.hail().expr.ir.AggOpRegistry.getType(j_agg_sig))
```

Reply:
Can't we just put the type on the ApplyScanOp IR node in Python?

Reply:
yeah, I think so. Probably can be added when the node is constructed with _agg_func().